src/main/cpp/src/SparkResourceAdaptorJni.cpp

```diff
@@ -1696,127 +1690,55 @@ class spark_resource_adaptor final : public rmm::mr::device_memory_resource {
-                thread_priority current = t_state.priority();
-                if (!is_to_wake_set || to_wake < current) {
-                  to_wake        = current;
-                  is_to_wake_set = true;
-                }
-              }
-            } break;
-            default: break;
-          }
-        }
-        // 4. Wake up the BUFN thread if we should
-        if (is_to_wake_set) {
-          long const thread_id_to_wake = to_wake.get_thread_id();
-          if (thread_id_to_wake > 0) {
-            // Don't wake up yourself on a free. It is not adding more memory for this thread
-            // to use on a retry and we might need a split instead to break a deadlock
-            auto const this_id = static_cast<long>(pthread_self());
-            auto const thread  = threads.find(thread_id_to_wake);
-            if (thread != threads.end() && thread->first != this_id) {
-              switch (thread->second.state) {
-                case thread_state::THREAD_BUFN:
-                  transition(thread->second, thread_state::THREAD_RUNNING);
-                  thread->second.wake_condition->notify_all();
-                  break;
-                case thread_state::THREAD_BUFN_WAIT:
-                  transition(thread->second, thread_state::THREAD_RUNNING);
-                  // no need to notify anyone, we will just retry without blocking...
-                  break;
-                case thread_state::THREAD_BUFN_THROW:
-                  // This should really never happen, this is a temporary state that is here only
-                  // while the lock is held, but just in case we don't want to mess it up, or throw
-                  // an exception.
-                  break;
-                default: {
-                  std::stringstream ss;
-                  ss << "internal error expected to only wake up blocked threads "
-                     << thread_id_to_wake << " " << as_str(thread->second.state);
-                  throw std::runtime_error(ss.str());
-                }
-              }
-            }
-          }
-        }
-      }
+    if (thread_to_wake != nullptr) {
+      transition(*thread_to_wake, thread_state::THREAD_RUNNING);
+      thread_to_wake->wake_condition->notify_all();
     }
   }
 
-  bool is_thread_bufn_or_above(JNIEnv* env, full_thread_state const& state)
+  /**
+   * Returns a boolean indicating if the thread is in BUFN (Blocked Until
+   * Further Notice) state or above. A thread is considered BUFN or above if:
+   * - It has pool_blocked set to true, OR
+   * - Its state is THREAD_BUFN, OR
+   * - Its thread ID is in the java_blocked_thread_ids set if it is provided.
+   *
+   * Threads in THREAD_BLOCKED state are NOT considered BUFN or above.
+   */
+  bool is_thread_bufn_or_above(
+    full_thread_state const& state,
+    std::optional<std::unordered_set<long>> const& java_blocked_thread_ids)
   {
+    LOG_INFO("is_thread_bufn_or_above: state: {}, java_blocked_thread_ids: {}",
```

Comment:
nit: Generally we have always had the logging follow a CSV/TSV like structure so that we can parse it and do post processing/analytics on it. Is the needed? Could this be moved to more of that format?